pkg/clients/redis/redis.go

```diff
@@ -0,0 +1,226 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package redis
+
+import (
+	"context"
+	"errors"
+	"strconv"
+	"time"
+
+	"github.com/crossplane/crossplane-runtime/pkg/password"
+
+	sdkerrors "github.com/aliyun/alibaba-cloud-sdk-go/sdk/errors"
+
+	aliredis "github.com/aliyun/alibaba-cloud-sdk-go/services/r-kvstore"
+
+	databasev1alpha1 "github.com/crossplane/provider-alibaba/apis/database/v1alpha1"
+	redisv1alpha1 "github.com/crossplane/provider-alibaba/apis/database/v1alpha1/redis"
+)
+
+var (
+	// ErrDBInstanceNotFound indicates DBInstance not found
+	ErrDBInstanceNotFound = errors.New("DBInstanceNotFound")
+	// ErrCodeInstanceNotFound error code of ServerError when DBInstance not found
+	ErrCodeInstanceNotFound = "InvalidDBInstanceId.NotFound"
+
+	defaultReadTime = 60 * time.Second
+	httpsScheme     = "https"
+)
+
+// Client defines Redis client operations
+type Client interface {
+	DescribeDBInstance(id string) (*DBInstance, error)
+	CreateAccount(id, username, password string) error
+	CreateDBInstance(*CreateRedisInstanceRequest) (*DBInstance, error)
+	DeleteDBInstance(id string) error
+	AllocateInstancePublicConnection(id string, port int) (string, error)
+	ModifyDBInstanceConnectionString(id string, port int) (string, error)
+}
+
+// DBInstance defines the DB instance information
+type DBInstance struct {
+	// Instance ID
+	ID string
+
+	// Instance status
+	Status string
+
+	// Endpoint specifies the connection endpoint.
+	Endpoint *databasev1alpha1.Endpoint
+}
+
+// CreateRedisInstanceRequest defines the request info to create DB Instance
+type CreateRedisInstanceRequest struct {
+	Name           string
+	Engine         string
+	EngineVersion  string
+	SecurityIPList string
+	InstanceClass  string
+	Password       string
+	Port           int
+	Config         string
+}
+
+type client struct {
+	redisCli *aliredis.Client
+}
+
+// NewClient creates new Redis RedisClient
+func NewClient(ctx context.Context, accessKeyID, accessKeySecret, region string) (Client, error) {
+	redisCli, err := aliredis.NewClientWithAccessKey(region, accessKeyID, accessKeySecret)
+	if err != nil {
+		return nil, err
+	}
+	c := &client{redisCli: redisCli}
+	return c, nil
+}
+
+func (c *client) DescribeDBInstance(id string) (*DBInstance, error) {
+	request := aliredis.CreateDescribeInstancesRequest()
+	request.Scheme = httpsScheme
+
+	request.InstanceIds = id
+
+	response, err := c.redisCli.DescribeInstances(request)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
It's better to wrap the error by `errors.Wrap(err, "xxx")`

Reply:
You can also refer the suggestion from @negz https://github.com/crossplane/provider-alibaba/pull/58#discussion_r595560525.